src/liballoc/collections/binary_heap.rs

```diff
@@ -1161,6 +1252,49 @@ impl<T> ExactSizeIterator for Drain<'_, T> {
 #[stable(feature = "fused", since = "1.26.0")]
 impl<T> FusedIterator for Drain<'_, T> {}
 
+/// A draining iterator over the elements of a `BinaryHeap`.
+///
+/// This `struct` is created by the [`drain_ordered`] method on [`BinaryHeap`]. See its
+/// documentation for more.
+///
+/// [`drain_ordered`]: struct.BinaryHeap.html#method.drain_ordered
+/// [`BinaryHeap`]: struct.BinaryHeap.html
+#[unstable(feature = "binary_heap_drain_ordered", issue = "59278")]
+#[derive(Debug)]
+pub struct DrainOrdered<'a, T> {
+    inner: &'a mut BinaryHeap<T>,
+}
+
+#[unstable(feature = "binary_heap_drain_ordered", issue = "59278")]
+impl<'a, T> Drop for DrainOrdered<'a, T> {
+    /// Removes heap elements in arbitrary order for efficiency.
```

Comment:
I think it should clear them in order, it seems surprising otherwise. The common theme of the `.drain()` iterators are that they have sideeffectful drops, so it wouldn't be out of place? It can be guarded by a needs-drop check?

Reply:
That has complexity side-effects I'd consider surprising, though -- I'd hope that `.drain().take(k)` is `O(k lg n)`, but if they have to drop in order, it'd be `O(n lg n)`.

The `BinaryHeap` itself drops in arbitrary order, so I don't think it's crazy for `drain` to also do so.  If someone really wants to drop in sorted order, then can always `.for_each(drop)`.

Reply:
In one way it seems to me that `.drain_ordered()` already is an intention of taking all the elements and dropping them in order. But I think your expectation there is fair.

Reply:
It looks like expensive operation on implicit call (such as `drop()`) is not a good thing in the Rust language. 
Therefore, I kept the behavior requiring user to explicitly consume elements to perform expensive `O(n ln n)` operation.

Reply:
Now `.drain_sorted()` is changed to always remove elements in heap order as @bluss has suggested.  
I suppose it simplifies usages a lot.